File: packages/php/tests/EnvelopeFixturesTest.php
<?php

declare(strict_types=1);

namespace Xyp\Tests;

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;
use Xyp\Auth;
use Xyp\Internal\Envelope;

/**
 * The fixtures are the Python SDK's output, which is verified against zeep (the
 * SOAP library the known-working XYP clients use) for every operation in
 * spec/wsdl. Same input, same bytes.
 */
final class EnvelopeFixturesTest extends TestCase
{
    /** The WSDLs we hold cover this many operations; fewer means the export went wrong. */
    private const MINIMUM_FIXTURES = 160;

    public function testTheFixtureFileHoldsEveryOperation(): void
    {
        self::assertGreaterThan(self::MINIMUM_FIXTURES, \count(self::fixtures()));
    }

    /**
     * @param array<string, mixed> $params
     */
    #[DataProvider('fixtureCases')]
    public function testEnvelopeIsByteIdenticalToTheZeepVerifiedFixture(
        string $operation,
        string $namespace,
        array $params,
        ?Auth $citizen,
        ?Auth $operator,
        string $envelope,
    ): void {
        self::assertSame($envelope, Envelope::build($operation, $namespace, $params, $citizen, $operator));
    }

    /**
     * @return iterable<string, array{string, string, array<string, mixed>, ?Auth, ?Auth, string}>
     */
    public static function fixtureCases(): iterable
    {
        foreach (self::fixtures() as $fixture) {
            $params = [];
            // A PHP array keeps paramOrder, the schema order the service expects.
            foreach ($fixture['paramOrder'] as $name) {
                self::assertArrayHasKey($name, $fixture['params'], "paramOrder names {$name}, which params does not hold");
                $value = $fixture['params'][$name];
                if (\in_array($name, $fixture['dateFields'], true)) {
                    self::assertIsString($value);
                    $value = new \DateTimeImmutable($value);
                }
                $params[$name] = $value;
            }
            $citizen = $operator = null;
            if ($fixture['auth'] !== null) {
                $citizen = Auth::otp($fixture['auth']['citizen']['regnum'], $fixture['auth']['citizen']['otp']);
                $fingerprint = base64_decode($fixture['auth']['operator']['fingerprintBase64'], true);
                self::assertIsString($fingerprint);
                $operator = Auth::fingerprint($fixture['auth']['operator']['regnum'], $fingerprint);
            }

            yield $fixture['operation'] => [$fixture['operation'], $fixture['namespace'], $params, $citizen, $operator, $fixture['envelope']];
        }
    }

    /**
     * @return list<array{
     *     operation: string,
     *     namespace: string,
     *     params: array<string, mixed>,
     *     paramOrder: list<string>,
     *     dateFields: list<string>,
     *     auth: ?array{citizen: array{regnum: string, otp: int}, operator: array{regnum: string, fingerprintBase64: string}},
     *     envelope: string,
     * }>
     */
    private static function fixtures(): array
    {
        $payload = file_get_contents(__DIR__ . '/fixtures/envelopes.json');
        self::assertIsString($payload);
        $fixtures = json_decode($payload, true, 512, JSON_THROW_ON_ERROR);
        self::assertIsArray($fixtures);

        /** @var list<array{operation: string, namespace: string, params: array<string, mixed>, paramOrder: list<string>, dateFields: list<string>, auth: ?array{citizen: array{regnum: string, otp: int}, operator: array{regnum: string, fingerprintBase64: string}}, envelope: string}> $fixtures */
        return $fixtures;
    }
}

File: packages/java/src/test/java/io/github/baljinnyamday/xyp/EnvelopeFixturesTest.java
package io.github.baljinnyamday.xyp;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.io.InputStream;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * The fixtures are the Python SDK's output, which is verified against zeep (the SOAP library the
 * known-working XYP clients use) for every operation in spec/wsdl. Same input, same bytes.
 */
class EnvelopeFixturesTest {

  /**
   * The WSDLs we hold cover this many operations; a fixture file with fewer than that is a sign the
   * export went wrong.
   */
  private static final int MINIMUM_FIXTURES = 160;

  @Test
  void theFixtureFileCoversEveryOperationInTheWsdls() throws IOException {
    assertTrue(
        fixtures().size() > MINIMUM_FIXTURES,
        "only " + fixtures().size() + " fixtures, want more than " + MINIMUM_FIXTURES);
  }

  @TestFactory
  Stream<DynamicTest> envelopesAreByteIdenticalToTheZeepVerifiedFixtures() throws IOException {
    return fixtures().stream()
        .map(
            fixture ->
                DynamicTest.dynamicTest(
                    fixture.get("operation").asText(),
                    () -> {
                      Auth citizen = null;
                      Auth operator = null;
                      JsonNode auth = fixture.get("auth");
                      if (!auth.isNull()) {
                        citizen = citizenAuth(auth.get("citizen"));
                        operator = operatorAuth(auth.get("operator"));
                      }
                      String envelope =
                          Envelope.build(
                              fixture.get("operation").asText(),
                              fixture.get("namespace").asText(),
                              params(fixture),
                              citizen,
                              operator);
                      assertEquals(fixture.get("envelope").asText(), envelope);
                    }));
  }

  /** Rebuilds the request in schema order, which the fixture keeps apart from the values. */
  private static Params params(JsonNode fixture) {
    Set<String> dateFields = new HashSet<>();
    fixture.get("dateFields").forEach(name -> dateFields.add(name.asText()));
    Params.Builder params = Params.builder();
    for (JsonNode nameNode : fixture.get("paramOrder")) {
      String name = nameNode.asText();
      JsonNode value = fixture.get("params").get(name);
      if (value == null) {
        throw new AssertionError("paramOrder names " + name + ", which params does not hold");
      }
      params.add(
          name, dateFields.contains(name) ? OffsetDateTime.parse(value.asText()) : value(value));
    }
    return params.build();
  }

  private static Object value(JsonNode value) {
    if (value.isBoolean()) {
      return value.booleanValue();
    }
    if (value.isIntegralNumber()) {
      // A long, so it is written without a ".0".
      return value.longValue();
    }
    if (value.isTextual()) {
      return value.textValue();
    }
    throw new AssertionError("unexpected fixture value " + value.getNodeType());
  }

  private static Auth citizenAuth(JsonNode citizen) {
    Auth auth = Auth.otp(citizen.get("regnum").asText(), citizen.get("otp").asInt());
    assertEquals(citizen.get("authType").asInt(), auth.authType().code());
    return auth;
  }

  private static Auth operatorAuth(JsonNode operator) {
    byte[] image = Base64.getDecoder().decode(operator.get("fingerprintBase64").asText());
    Auth auth = Auth.fingerprint(operator.get("regnum").asText(), image);
    assertEquals(operator.get("authType").asInt(), auth.authType().code());
    return auth;
  }

  private static List<JsonNode> fixtures() throws IOException {
    try (InputStream in = EnvelopeFixturesTest.class.getResourceAsStream("/envelopes.json")) {
      if (in == null) {
        throw new AssertionError("src/test/resources/envelopes.json is missing");
      }
      List<JsonNode> fixtures = new ArrayList<>();
      new ObjectMapper().readTree(in).forEach(fixtures::add);
      return fixtures;
    }
  }
}
